packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/macro_expand.test.rs

```diff
@@ -0,0 +1,240 @@
+use assert_json_diff::assert_json_eq;
+
+use super::*;
+use crate::test_utils::*;
+
+#[test]
+fn macros_get_expanded_to_their_referents() {
+    let dast_root = dast_root_no_position(r#"<point name="p" />$p"#);
+    let mut flat_root = FlatRoot::from_dast(&dast_root);
+    Expander::expand(&mut flat_root);
+    assert_json_eq!(
+        serde_json::to_value(&flat_root).unwrap(),
+        json!(
+            {
+                "type": "FlatRoot",
+                "children": [0],
+                "nodes": [
+                  {
+                    "type": "Element",
+                    "name": "document",
+                    "children": [1, 2],
+                    "attributes": [],
+                    "idx": 0
+                  },
+                  {
+                    "type": "Element",
+                    "name": "point",
+                    "parent": 0,
+                    "children": [],
+                    "attributes": [
+                      {
+                        "name": "name",
+                        "parent": 1,
+                        "children": ["p"]
+                      }
+                    ],
+                    "idx": 1
+                  },
+                  {
+                    "type": "Element",
+                    "name": "point",
+                    "parent": 0,
+                    "children": [],
+                    "attributes": [],
+                    "idx": 2,
+                    "extending": {
+                      "node_idx": 1,
+                      "unresolved_path": null
+                    }
+                  }
+                ]
+              }
+        )
+    );
+
+    let dast_root = dast_root_no_position(
+        &r#"<a name="x">
+            <b name="y">
+                <c name="z" />
+                $x.y
+            </b>
+            <f>
+                <e name="w" />
+            </f>
+        </a>
+        $w
+        <d name="y" />"#
+            .replace("\n", "")
+            .replace("  ", ""),
+    );
+    let mut flat_root = FlatRoot::from_dast(&dast_root);
+    Expander::expand(&mut flat_root);
+    assert_eq!(
+        r#"<document><a name="x"><b name="y"><c name="z" /><b /></b><f><e name="w" /></f></a><e /><d name="y" /></document>"#,
+        flat_root.to_xml()
```

Comment:
What is the desired behavior of the `.to_xml()` function? It is supposed to encode what core will try to do? Because, core copies the children when extending a component (as opposed to a state variable). Should the `.to_xml()` indicate that the second `<b />` actually will end up getting children?

(In this example, core will run into a circular dependence, as the children of the `<b />` created from the reference will contain a copy of itself. It currently hangs the browser as we don't have circular dependence checks.)

Reply:
Oh, it looks like you have more macro resolve tests, below.

Reply:
`to_xml` should only be used for testing. It's easier to see what some tests are trying to do in XML form rather than looking at bit data structures. I moved it to `test_utils` so that it cannot be invoked from core.